gobblin-runtime/src/main/java/gobblin/runtime/JobContext.java

```diff
@@ -170,38 +169,131 @@ public boolean isJobLockEnabled() {
+   * Set final {@link JobState} of the given job.
    */
-  public Optional<JobHistoryStore> getJobHistoryStoreOptional() {
-    return this.jobHistoryStoreOptional;
+  void setFinalJobState() {
+    this.jobState.setEndTime(System.currentTimeMillis());
+    this.jobState.setDuration(this.jobState.getEndTime() - this.jobState.getStartTime());
+
+    for (TaskState taskState : this.jobState.getTaskStates()) {
+      // Set fork.branches explicitly here so the rest job flow can pick it up
+      this.jobState.setProp(ConfigurationKeys.FORK_BRANCHES_KEY,
+          taskState.getPropAsInt(ConfigurationKeys.FORK_BRANCHES_KEY, 1));
+    }
+
+    if (this.jobState.getState() == JobState.RunningState.SUCCESSFUL) {
+      // Reset the failure count if the job successfully completed
+      this.jobState.setProp(ConfigurationKeys.JOB_FAILURES_KEY, 0);
+    }
+
+    if (this.jobState.getState() == JobState.RunningState.FAILED) {
+      // Increment the failure count by 1 if the job failed
+      int failures = this.jobState.getPropAsInt(ConfigurationKeys.JOB_FAILURES_KEY, 0) + 1;
+      this.jobState.setProp(ConfigurationKeys.JOB_FAILURES_KEY, failures);
+    }
+  }
+
+  /**
+   * Commit the job on a per-dataset basis.
+   */
+  void commit() throws Exception {
+    Map<String, JobState.DatasetState> datasetStatesByUrns = this.jobState.getDatasetStatesByUrns();
+    for (Map.Entry<String, JobState.DatasetState> entry : datasetStatesByUrns.entrySet()) {
+      setFinalDatasetState(entry.getValue());
+      if (canCommitDataset(this.jobCommitPolicy, entry.getValue())) {
+        try {
+          commitDataset(entry.getValue());
+        } finally {
+          persistDatasetState(entry.getKey(), entry.getValue());
+        }
+      }
+    }
+  }
+
+  private void setFinalDatasetState(JobState.DatasetState datasetState) {
+    for (TaskState taskState : datasetState.getTaskStates()) {
+      if (taskState.getWorkingState() != WorkUnitState.WorkingState.SUCCESSFUL) {
+        // The dataset state is set to FAILED if any task failed
+        datasetState.setState(JobState.RunningState.FAILED);
+        return;
+      }
+    }
+
+    datasetState.setState(JobState.RunningState.SUCCESSFUL);
+  }
+
+  /**
+   * Check if it is OK to commit the output data of a dataset.
+   */
+  private boolean canCommitDataset(JobCommitPolicy commitPolicy, JobState.DatasetState datasetState) {
+    // Only commit a dataset if 1) COMMIT_ON_PARTIAL_SUCCESS is used, or 2)
+    // COMMIT_ON_FULL_SUCCESS is used and all of the tasks of the dataset have succeeded.
+    return commitPolicy == JobCommitPolicy.COMMIT_ON_PARTIAL_SUCCESS
+        || (commitPolicy == JobCommitPolicy.COMMIT_ON_FULL_SUCCESS
+        && datasetState.getState() == JobState.RunningState.SUCCESSFUL);
+  }
+
+  /**
+   * Commit the output data of a dataset.
+   */
+  @SuppressWarnings("unchecked")
+  private void commitDataset(JobState.DatasetState datasetState) throws Exception {
```

Comment:
Can we use a more concrete exception class like IOException? Using Exception kind of defeats the purpose of static exception checking.